### DISCUSSION TITLE
Function->Configuration VERSUS appproperties.json?

### DISCUSSION THREAD

**Discussion by rfreytag** [n3wt0n/AzureWebAppSSLManager#41] [Q&A]
Some values like CertificatePassword are part of the Function->Configuration.  

Other values like "Hostname" are part of appproperties.json.  

Both seem similarly durable or changeable.  You might want to change the Hostname on the function.  Why or why not?

What defines which values should be preserved where?

**Comment by n3wt0n** [comment]
In `Configuration` you have params and settings that are relative to the Azure Function that runs the AzureWebAppSSLManager logic and that are needed for the software to work.

`appproperties.json` instead contains the bits the software has to work on, meaning the information about the targets where to install/renew the certificates. The _Hostname_ you mention is the hostname for the target webapp/function that the software has to install the certificate on, and it is needed by the logic to identify and retrieve the target object

**Comment by rfreytag** [comment]
Thanks for the reply Peter.  

Both Configuration and appproperties.json are consumed by AzureWebAppSSLManager and with only a few values sent on to SendGrid by the Azure function - right?

As I use AzureWebAppSSLManager I find myself wanting to change the Hostname easily.  Changing a Function->Configuraton setting is faster and easier than editing appproperties.json and then having to kick off the Function.  

I would like if AzureWebAppSSLManager->Configuration had a "apppropertiesPrimaryLocation" = True/False.  

If True then look first in appproperties.json for configuration, and if False look in Function->Configuration for configuration info.

Any value not found in the first location would be sought in the alternative location.  

Missing configuration/property values would then throw an error.

This would keep the original behavior if "appparopertiesPrimaryLocation=True" AND improve on-line usability.

n3wt0n - I am proposing to do this - but I would like to know it would be acceptable to you.

Thanks,
Richard Freytag




-----Original Message-----

**Reply by n3wt0n** [reply]
Ideally you shouldn't have to change the hostname very often... the json file can contain an array of properties, meaning that if for example you need to install/renew certificates for 3 Azure WebApps, you would have all 3 in there (with the 3 different hostnames, etc.)

What is your use case for having to change the hostname of the WebApp you target?

**Comment by rfreytag** [comment]
I am relatively new so it isn't clear to me how to configure AzureWebAppSSLManager to get mutliple Let's Encrypt certificates.

While I'm at it - is there a way to have *.pfx files automatically imported as an SSL certificate into a WebJob?

As I say - I am new with lots of probably simple questions.

Thank you,
Richard F.


-----Original Message-----

**Reply by n3wt0n** [reply]
**Multiple certificates**

If you want _AzureWebAppSSLManager _ to create and install multiple certificates, all you have to do is adding more "sections" in the app properties.

This, for example, will give you a single certificate:

```json
[
  {
    "Hostname": "SUB.YOUR_DOMAIN.EXT",
    "ResourceName": "WEBAPP_RESOURCE_NAME",
    "ResourceResGroup": "RESOURCE_GROUP_CONTAINING_THE_WEBAPP",
    "AzureDnsZoneName": "DNS_RESOURCE_NAME",
    "AzureDnsResGroup": "RESOURCE_GROUP_CONTAINING_THE_DNS"
  }
]
```

This, instead, 2 certs for the _SUB1_ and _SUB2_ subdomains

```json
[
  {
    "Hostname": "SUB1.YOUR_DOMAIN.EXT", 
    "ResourceName": "WEBAPP_RESOURCE_NAME",
    "ResourceResGroup": "RESOURCE_GROUP_CONTAINING_THE_WEBAPP",
    "AzureDnsZoneName": "DNS_RESOURCE_NAME",
    "AzureDnsResGroup": "RESOURCE_GROUP_CONTAINING_THE_DNS"
  },
 {
    "Hostname": "SUB2.YOUR_DOMAIN.EXT",
    "ResourceName": "WEBAPP_RESOURCE_NAME",
    "ResourceResGroup": "RESOURCE_GROUP_CONTAINING_THE_WEBAPP",
    "AzureDnsZoneName": "DNS_RESOURCE_NAME",
    "AzureDnsResGroup": "RESOURCE_GROUP_CONTAINING_THE_DNS"
  }
]
```
